scripts/swap-beacon/src/adapter/odos.ts

```diff
@@ -0,0 +1,171 @@
+import { Adapter, ExtendedTargetCall, TokenToSend } from "../types";
+import { Address, Hex } from "viem";
+import { TokenAmount } from "@pythnetwork/express-relay-evm-js";
+import axios from "axios";
+import { getSwapAdapterConfig } from "../index";
+import { ODOS_TOKEN_MAP } from "./const";
+
+export class OdosAdapter implements Adapter {
+  chainIds: string[] = ["mode"];
+  tokenMap = this.constructTokenMap();
+
+  private constructTokenMap(): Record<string, Record<string, any>> {
+    return Object.keys(ODOS_TOKEN_MAP).reduce(
+      (res, key) => ({
+        ...res,
+        key: Object.fromEntries(
+          Object.entries(ODOS_TOKEN_MAP[key]).map(([k, v]) => [
+            k.toLowerCase(),
+            v,
+          ])
+        ),
+      }),
+      {}
+    );
+  }
+
+  private getTokenMap(chainId: string, tokenIn: Address, tokenOut: Address) {
+    const tokenMap = this.tokenMap[chainId];
+    if (!tokenMap) {
+      throw new Error("Chain Id not supported");
+    }
+    if (!tokenMap[tokenIn.toLowerCase()]) {
+      throw new Error("Token In not supported");
+    }
+    if (!tokenMap[tokenOut.toLowerCase()]) {
+      throw new Error("Token Out not supported");
+    }
+    return tokenMap;
+  }
+
+  private async getPrice(chainId: string, token: Address): Promise<number> {
+    const swapAdapterConfig = getSwapAdapterConfig(chainId);
+    const response = await axios.get(
+      `https://api.odos.xyz/pricing/token/${swapAdapterConfig.chainIdNum}/${token}`
+    );
+    return response.data.price as number;
+  }
+
+  private async estimateAmountIn(
+    chainId: string,
+    tokenIn: Address,
+    tokenOut: Address,
+    amountOut: bigint
+  ) {
+    const tokenMap = this.getTokenMap(chainId, tokenIn, tokenOut);
+    const decimalsIn = tokenMap[tokenIn.toLowerCase()].decimals as number;
+    const decimalsOut = tokenMap[tokenOut.toLowerCase()].decimals as number;
+
+    const [priceIn, priceOut] = await Promise.all([
+      this.getPrice(chainId, tokenIn),
+      this.getPrice(chainId, tokenOut),
+    ]);
+
+    return BigInt(
+      Math.ceil(
+        ((((Number(amountOut) * priceOut) / priceIn) * 10 ** decimalsIn) /
```

Comment:
I find this formula a bit unintuitive. maybe something like this is easier to understand?

`amountOut * (priceOut/(10**decimalsOut) / (priceIn/(10**decimalsIn))`
And then we can store `(priceOut/(10**decimalsOut) / (priceIn/(10**decimalsIn))` as conversion rate?